### PR TITLE
Revert changes in logit size for semantic segmentation models

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR reverts the changes made in the size of the returned logits of semantic segmentation models made in #15469 because it leads in some situations to two consecutive resizes:
- one to resize the output to the same size of the input
- one to resize the output to the size of the original image
Two different interpolations with nearest neighbors is a bad idea and users will get a best result by just doing one interpolation to get the segmentation mask to the size they want.

This restores backward compatibility, but still introduces a new argument `resize_logits` for situations where the user wants the logits the same size as the `pixel_values` passed to the model.

### CODE REVIEW

File: tests/test_modeling_beit.py [hunk 1]
```diff
@@ -162,11 +162,11 @@ def create_and_check_for_image_segmentation(self, config, pixel_values, labels,
         model.eval()
         result = model(pixel_values)
         self.parent.assertEqual(
-            result.logits.shape, (self.batch_size, self.num_labels, self.image_size, self.image_size)
+            result.logits.shape, (self.batch_size, self.num_labels, self.image_size * 2, self.image_size * 2)
```
Comment (sgugger) [inline tests/test_modeling_beit.py:165]: This is really weird @NielsRogge . Why does BeIT return a bigger size with image_size=30?

File: src/transformers/models/beit/configuration_beit.py [hunk 2]
```diff
@@ -93,10 +93,8 @@ class BeitConfig(PretrainedConfig):
             Whether to concatenate the output of the auxiliary head with the input before the classification layer.
         semantic_loss_ignore_index (`int`, *optional*, defaults to 255):
             The index that is ignored by the loss function of the semantic segmentation model.
-        legacy_output (`bool`, *optional*, defaults to `False`):
-            Whether to return the legacy outputs or not (with logits of shape `height / 4 , width / 4`)
-
-            This argument is only present for backward compatibility reasons and will be removed in v5 of Transformers.
+        resize_logits (`bool`, *optional*, defaults to `False`):
+            Whether to resize the logits to the same size as the `pixel_values` or not.
```
Comment (NielsRogge) [inline src/transformers/models/beit/configuration_beit.py:97]: I'm not sure if this is useful.

Why would people want to resize the logits to the size of the `pixel_values` (which is typically something like 512x512 or 640x640 i.e. square)? All they care about is resize the logits to the size of the original image.

Comment (NielsRogge) [reply src/transformers/models/beit/configuration_beit.py:97]: Yes but if they are using a custom loss, they can interpolate themselves if they want?

Comment (sgugger) [reply src/transformers/models/beit/configuration_beit.py:97]: They can also write the whole model themselves if they want. The goal of the library is to provide them with things to make it easier ;-)

Comment (NielsRogge) [reply src/transformers/models/beit/configuration_beit.py:97]: Yes but I wouldn't add this in the same spirit as why we don't add custom losses. If people want to use this model with a custom loss, they can easily use our model implementation to get the `logits`, then do whatever they want with it.

They don't need to re-implement the whole model for that.

Comment (FrancescoSaverioZuppichini) [reply src/transformers/models/beit/configuration_beit.py:97]: I agree with @NielsRogge , the only use case I can come up with is a custom loss, and usually, the interpolation is done in the loss directly. Moreover, if we had the ability to resize the logits we should also be sure the user can decide *how* the logits are interpolated ("bilinear" vs something else)

Comment (NielsRogge) [reply src/transformers/models/beit/configuration_beit.py:97]: To expand my point, suppose people want to get the `logits` the same size as the `pixel_values`. Here's how they can achieve that without having to override the entire model:
```
import transformers import SegformerPreTrainedModel, SegformerForSemanticSegmentation
import torch.nn as nn

class CustomSegFormer(SegformerPreTrainedModel):
    def __init__(self, config):
        super().__init__(config)
        self.segformer = SegformerForSemanticSegmentation.from_pretrained("...")

    def forward(self, pixel_values, labels=None):
        # forward to get raw logits
        outputs = self.segofmrer(ids, attention_mask=mask, token_type_ids=token_type_ids)
        logits = outputs.logits
        
        # next, do whatever they want with the logits to compute loss
        loss = None
        if labels is not None:
               # upsample logits here
               # calculate custom loss here

        return output
```

Comment (LysandreJik) [reply src/transformers/models/beit/configuration_beit.py:97]: Not a vision engineer so probably not seeing every use case, but I'd favor minimal changes to the output of the model and to let users do what they wish with the output of the model. Resizing logits so that they can be fed to a custom loss of theirs doesn't look like it adds too much complexity and they'll already be computing the loss outside of the model.

Comment (sgugger) [reply src/transformers/models/beit/configuration_beit.py]: Removed that bit!

**Review by LysandreJik — APPROVED** [review]
LGTM, thank you @sgugger!

**Review by FrancescoSaverioZuppichini — COMMENTED** [review]
LGTM, thanks @sgugger !
